Client Credentials Request

msya opened the pull request:
This PR implements the auth request `/oauth/authorize/client`. 

Here is the implementation:

#### Authenticator

Authenticator is an interface the client will use to perform any type of authentication. Basic authentication with a client id and client secret can be done by using the `clientCredentials` method  in the interface.

```
interface Authenticator {

    fun clientCredentials(authCallback: AuthCallback)

}
```

#### Creating Authenticator

```
interface Authenticator {

    fun clientCredentials(authCallback: AuthCallback)

    companion Factory {
 
         fun create(serverConfig: ServerConfig)

    }
}
```

The `Authenticator` interface contains a factory to create an instance of it by supplying a `ServerConfig` object.

#### VimeoCallAdapter

If an error occurs, the API will send an error code and developer message. The information needs to be parsed into an `ApiError` object. This call adapter was created in order to do this.

#### Usage

This is how it looks from a client's perspective.

```
            val serverConfig = ServerConfig(clientId, clientSecret, scopes)
            val authenticator = Authenticator.Factory.create(serverConfig)

           authenticator.clientCredentials(object : AuthCallback {

                override fun onSuccess(authResponse: ApiResponse.Success<String>) {
                    
                }

                override fun onGenericError(genericFailure: ApiResponse.Failure.GenericFailure) {
                    
                }

                override fun onApiError(apiFailure: ApiResponse.Failure.ApiFailure) {
                    
                }

                override fun onExceptionError(exceptionFailure: ApiResponse.Failure.ExceptionFailure) {
                    
                }
            })

```

`VimeoAccount` is not provided to the user only the access token is provided. This is because that's only important needed to make subsequent requests. Other info on the `VimeoAcccount` could be retrieved from the API such as the user object.

brentwatson commented on api/src/main/java/com/vimeo/networking2/requests/ApiRequestRetrofitCallback.kt:
```diff
@@ -0,0 +1,34 @@
+        if (response.isSuccessful) {
+            response.body()?.let {
+                apiResponseCallback.onSuccess(it)
+            } ?: apiResponseCallback.onError(ApiError(errorMessage = "Failed to parse data."))
```
Should this error message be that body is null instead of assuming that parsing failed?

msya replied:
I updated it to "Response body is null" to be more specific.

brentwatson commented on api/src/main/java/com/vimeo/networking2/requests/ApiResponseCallback.kt:
```diff
@@ -0,0 +1,15 @@
+import com.vimeo.networking2.ApiError
+
+/**
+ * Informs you of the result of an API request. Provides the response or
```
minor / optional: docs don't need to wrap.

brentwatson commented on api/src/main/java/com/vimeo/networking2/utils/ApiResponseErrorConverter.kt:
```diff
@@ -0,0 +1,26 @@
+ *
+ * @param retrofit Required for using the response body converter to do the conversion.
+ */
+class ApiResponseErrorConverter(private val retrofit: Retrofit) {
```
I'd suggest you extract an interface for this class and change `ApiRequestRetrofitCallback` and `Authenticator` to take the interface. That allows those classes to be more easily tested.

brentwatson commented on api/src/main/java/com/vimeo/networking2/utils/ApiResponseErrorConverter.kt:
```diff
@@ -0,0 +1,26 @@
+            )
+            response.errorBody()?.let { errorConverter.convert(it) } ?: ApiError()
+        } catch (e: Exception) {
+            ApiError()
```
Should we do anything else here? Maybe log the exception or put something from the exception into `ApiError` so that it is surfaced?

msya replied:
I added the exception message to `ApiError`.

brentwatson commented on api/src/main/java/com/vimeo/networking2/utils/ApiResponseErrorConverter.kt:
```diff
@@ -0,0 +1,26 @@
+        return try {
+            val errorConverter = retrofit.responseBodyConverter<ApiError>(
+                ApiError::class.java,
+                arrayOfNulls<Annotation>(0)
```
optional: this can just be `emptyArray()`.

brentwatson commented on api/src/main/java/com/vimeo/networking2/utils/ApiResponseErrorConverter.kt:
```diff
@@ -0,0 +1,26 @@
+                ApiError::class.java,
+                arrayOfNulls<Annotation>(0)
+            )
+            response.errorBody()?.let { errorConverter.convert(it) } ?: ApiError()
```
Do we want to include that the fact that the error body is blank somewhere in `ApiError`?

Both of the empty ApiError instances should probably at least include the http error code.

brentwatson replied:
Optional: you could use the new `ifEmpty { }` lambda here to provide a default value instead of using the elvis operator.

brentwatson commented on auth/src/main/java/com/vimeo/networking2/Authenticator.kt:
```diff
@@ -31,31 +29,23 @@ class Authenticator(private val serverConfig: ServerConfig) {
      */
     private var retrofitServicesCache: RetrofitServicesCache
 
+    private var apiErrorResponseErrorConverter: ApiResponseErrorConverter
```
Other 2 member vars have kdocs. Maybe add the for this var too.

brentwatson commented on auth/src/main/java/com/vimeo/networking2/Authenticator.kt:
```diff
@@ -31,31 +29,23 @@ class Authenticator(private val serverConfig: ServerConfig) {
     init {
         retrofit = RetrofitSetupComponent.retrofit(serverConfig)
         retrofitServicesCache = RetrofitSetupComponent.retrofitCache(retrofit)
+        apiErrorResponseErrorConverter = RetrofitSetupComponent.apiResponseErrorConverter(retrofit)
```
I don't think the `init` block is actually needed. I think you could simply init this in-line.

brentwatson commented on auth/src/main/java/com/vimeo/networking2/Authenticator.kt:
```diff
@@ -31,31 +29,23 @@ class Authenticator(private val serverConfig: ServerConfig) {
-        val call = apiService.authorizeWithClientCredentialsGrant(
+    fun clientCredentials(): ClientCredentialsAuthenticator {
+        return ClientCredentialsInteractor(
+            apiErrorResponseErrorConverter,
```
Optional: I suggest added named args here. Makes it a bit nicer to read.

brentwatson commented on auth/src/main/java/com/vimeo/networking2/requests/clientcredentials/ClientCredentialsInteractor.kt:
```diff
@@ -0,0 +1,52 @@
+    private val authService: AuthService,
+    private val authHeaders: String,
+    private val scopes: String,
+    private val setAuthTokenCallback: (String) -> Unit
```
This val should probably just be named `authTokenCallback` not `setAuthTokenCallback`.

brentwatson commented on auth/src/main/java/com/vimeo/networking2/Authenticator.kt:
```diff
@@ -31,31 +29,23 @@ class Authenticator(private val serverConfig: ServerConfig) {
-        })
+            serverConfig.scopes.joinToString { it.value }
+        ) {
+            setAccessToken(it)
```
Optional (and maybe a personal pref) but I find prefer using last-arg lambads like this when something is being executed and we're expecting a response.

eg:
```kotlin
makeNetworkRequest(scope) { response ->
 ...
}
```

But for constructors it's weird because in a case like this where the lambda executed after a call to `authenticate`, and only on success, this makes it so that it's not obvious what the lambda is doing or when it's going to be called, so a dev needs to dig into the implementation to find out.

simple eg:
```kotlin
NetworkRequestor(scope, url, blah, blah) {
  8080
}
```
here you would need to go into the impl to find that this is some callback to get a port number, why it's needed, and when it's used.

If you agree with this, I'd suggest adding this lambda to the params instead of hanging it off the end.

i.e.:
```kotlin
return ClientCredentialsInteractor(
            apiResponseErrorConverter = apiErrorResponseErrorConverter,
            authService = retrofitServicesCache.getService(AuthService::class.java),
            authHeaders = RetrofitSetupComponent.authHeaders(serverConfig.clientId, serverConfig.clientSecret),
            scopes = serverConfig.scopes.joinToString { it.value },
            authTokenCallback = { setAccessToken(it) }
        )
```

... to me this makes things more clear when & why it will be used.

brentwatson commented on auth/src/main/java/com/vimeo/networking2/requests/AuthCallback.kt:
```diff
@@ -0,0 +1,26 @@
+interface AuthCallback {
+
+    /**
+     * The authentication was successful and the access token
```
nit: can be 1 line.

brentwatson commented on auth/src/main/java/com/vimeo/networking2/requests/clientcredentials/ClientCredentialsInteractor.kt:
```diff
@@ -0,0 +1,52 @@
+        val call = authService.authorizeWithClientCredentialsGrant(
+            authHeaders,
+            GrantType.CLIENT_CREDENTIALS.value,
+            scopes)
```
If I run our formatter on this file I get 4 changes.

brentwatson commented on auth/src/main/java/com/vimeo/networking2/requests/clientcredentials/ClientCredentialsInteractor.kt:
```diff
@@ -0,0 +1,52 @@
+        val apiResponseCallback = object : ApiResponseCallback<VimeoAccount> {
+
+            override fun onSuccess(response: VimeoAccount) {
+                response.accessToken?.let {
```
Do we want to call `onError` if the `accessToken` is null? Currently this would no-op.

msya commented:
@brentwatson @anthonycr I made the code review changes. I implemented a pattern to handle cancelling and making requests. The error response is now converted to a model in a different thread. Back to you.

anthonycr commented on api/src/main/java/com/vimeo/networking2/utils/CallExtensions.kt:
```diff
@@ -0,0 +1,29 @@
+ * function. When the request comes back, the method will resume
+ * the suspending method.
+ */
+suspend fun <T : Any> Call<T>.awaitResponse(): Response<T> =
```
since this requires coroutines, should it be in an extension module (we can come up with a name for the module) rather than the main api module

anthonycr commented on auth/src/main/java/com/vimeo/networking2/requests/clientcredentials/ClientCredentialsInteractor.kt:
```diff
@@ -0,0 +1,90 @@
+    private val authService: AuthService,
+    private val authHeaders: String,
+    private val apiScopes: List<ScopeType>,
+    private val coroutineScope: CoroutineScope,
```
Will java consumers be able to use this interactor? Perhaps we should have a separate `auth-coroutines` module

anthonycr commented on auth/src/main/java/com/vimeo/networking2/requests/clientcredentials/ClientCredentialsAuthenticator.kt:
```diff
@@ -0,0 +1,17 @@
+     *
+     * @param authCallback informs you of the result of the response.
+     */
+    fun authenticate(authCallback: AuthCallback)
```
Should the `authenticate` function return an `ApiRequest` rather than the whole authenticator itself implementing it?

anthonycr commented on api/src/main/java/com/vimeo/networking2/requests/ApiRequestRetrofitCallback.kt:
```diff
@@ -0,0 +1,39 @@
+ * @param errorResponseConverter Used to convert an error in the response to [ApiError].
+ * @param apiResponseCallback       Used to notify you of the response or error.
+ */
+class ApiRequestRetrofitCallback<ResponseType_T>(
```
this class is not used anywhere, will it be used in a later implementation?

anthonycr commented on api/src/main/java/com/vimeo/networking2/requests/ApiResponse.kt:
```diff
@@ -0,0 +1,40 @@
+     *
+     * @param accessToken The accessToken parsed by this successful response.
+     */
+    data class Success<out T>(val accessToken: T) : ApiResponse<T>()
```
this looks like a generic response enumeration, but the property is called `accessToken`, should it be called `data` or something similar?

anthonycr replied:
bump

anthonycr commented on api/src/main/java/com/vimeo/networking2/requests/ApiResponseCallback.kt:
```diff
@@ -0,0 +1,14 @@
+/**
+ * Informs you of the result of an API request. Provides the response or any errors.
+ */
+interface ApiResponseCallback<ResponseType_T> {
```
Should this be using the enumerated `ApiResponse` instead of having singular success and error methods?

msya commented:
@anthonycr I made the changes as we did in person. Back to you.

anthonycr commented on api/src/main/java/com/vimeo/networking2/adapters/ErrorHandlingCallAdapter.kt:
```diff
@@ -0,0 +1,33 @@
+ * Custom call adapter to handle errors.
+ *
+ * @param responseType              Type of the response.
+ * @param callbackExecutor          The executor upon which the request is made.
```
Is this documentation correct? Seems like it should be the executor on which the caller is notified.

anthonycr commented on api/src/main/java/com/vimeo/networking2/adapters/ErrorHandlingCallAdapterFactory.kt:
```diff
@@ -0,0 +1,35 @@
+    /**
+     * Creates and returns a [ErrorHandlingCallAdapter].
+     */
+    override fun get(returnType: Type,
```
minor formatting: the first parameter should be moved to a new line

anthonycr commented on api/src/main/java/com/vimeo/networking2/adapters/ErrorHandlingCallAdapterFactory.kt:
```diff
@@ -0,0 +1,35 @@
+            ApiError::class.java,
+            emptyArray()
+        )
+        return ErrorHandlingCallAdapter<Any>(responseType, callbackExecutor!!, errorResponseConverter)
```
I can't remember if we discussed this and maybe it's guaranteed non null elsewhere in this PR, but could it be null?

anthonycr commented on api/src/main/java/com/vimeo/networking2/adapters/VimeoCallAdapter.kt:
```diff
@@ -0,0 +1,76 @@
+     * Send response on [callbackExecutor].
+     */
+    fun sendResponse(action: () -> Unit) {
+        callbackExecutor.execute { action() }
```
`callbackExecutor.execute(action)` results in cleaner bytecode as it allows the `action` to be wrapped in a static `Runnable` instead of the anonymous instance that results from the current code. Going further, inlining `sendResponse eliminates the extra class creation all together.

anthonycr commented on api/src/main/java/com/vimeo/networking2/adapters/VimeoCallAdapter.kt:
```diff
@@ -0,0 +1,76 @@
+    /**
+     * Determine is the response has a body.
+     */
+    fun Response<T>.hasBody() = isSuccessful && body() != null
```
can be private

anthonycr commented on api/src/main/java/com/vimeo/networking2/adapters/VimeoCallAdapter.kt:
```diff
@@ -0,0 +1,76 @@
+    /**
+     * Parse the error body into a [ApiError] object.
+     */
+    fun Response<T>.parseApiError() = errorBody()?.let { responseBodyConverter.convert(it) }
```
can be private

anthonycr commented on api/src/main/java/com/vimeo/networking2/adapters/VimeoCallback.kt:
```diff
@@ -0,0 +1,32 @@
+     * Generic error occurred. This occurred was not caused by the API rather it was due to
+     * parsing the data or no response body was null.
+     */
+    fun onGenericError(requestCode: Int)
```
Should the parameter name be responseCode?

anthonycr commented on auth/src/main/java/com/vimeo/networking2/requests/AuthService.kt:
```diff
@@ -17,7 +17,7 @@ interface AuthService {
      *
      * @param authHeader It is created from the client id and client secret.
-     * @param grantType  Determines whether you have access to public or private data.
+     * @param grantType  Determines whether you have access to public or private accessToken.
```
This change looks to be the result of a broken rename and should be reverted.

anthonycr commented on auth/src/main/java/com/vimeo/networking2/utils/AuthExtensions.kt:
```diff
@@ -0,0 +1,47 @@
+package com.vimeo.networking2.utils
```
Should this file utilize `@file:JvmName` for better java interop or are we okay with the name being `AuthExtensions`?

msya replied:
This extension will only be used internally by the SDK. Therefore, I didn't add `@file:JvmName`. If public extensions are added to this class later on, then I could create a public extensions file and add `@file:JvmName`.

msya commented:
@anthonycr Thanks for the review. Back to you.

anthonycr commented on auth/src/main/java/com/vimeo/networking2/Authenticator.kt:
```diff
@@ -1,73 +1,68 @@
+ *
+ * Ex:
+ *
+ * val authenticator = Authenticator.create(serverConfig)
```
could surround this codeblock with ` ``` ` to get it formatted as code

anthonycr commented on auth/src/main/java/com/vimeo/networking2/Authenticator.kt:
```diff
@@ -1,73 +1,68 @@
+     * Factory to create an instance of [Authenticator].
      */
-    private var retrofitServicesCache: RetrofitServicesCache
+    companion object Factory {
```
elegant use of the companion object 💃

anthonycr commented on api/src/main/java/com/vimeo/networking2/adapters/VimeoCallback.kt:
```diff
@@ -0,0 +1,40 @@
+package com.vimeo.networking2.adapters
```
is the `adapters` package the correct package for this? could it be moved into the parent package as that will be more accessible for consumers. perhaps we could follow what some other libraries do where they have an `internal` package where we could put the internal classes in the adapter. happy to talk offline about the package structure.

anthonycr commented on auth/src/main/java/com/vimeo/networking2/requests/interactors/ClientCredentialsInteractor.kt:
```diff
@@ -0,0 +1,34 @@
+ * @param authHeaders   Client id and client secret headers.
+ * @param scopes        API scopes to support.
+ */
+internal class ClientCredentialsInteractor(
```
this class is so small, and it's also internal - could the `AuthenticatorImpl` just do what this class is doing? There's almost as much setup of the `ClientCredentialsInteractor` required as there is just to talk to the `AuthService` directly.
